src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -0,0 +1,148 @@
+            try
+            {
+                IDictionary<string, StringValues> headers = new Dictionary<string, StringValues>();
+                if (_options.IncludeRequestHeaders)
+                {
+                    IDictionary<string, StringValues> sanitizedHeaders = SanitizeRequestHeaders(httpContext.Request.Headers);
+                    if (sanitizedHeaders != null)
+                    {
+                        headers = sanitizedHeaders;
+                    }
+                }
+
+                IDictionary<string, StringValues> body = new Dictionary<string, StringValues>();
+                if (_options.IncludeRequestBody)
+                {
+                    IDictionary<string, StringValues> sanitizedBody = await SanitizeRequestBodyAsync(httpContext.Request.Body);
+                    if (sanitizedBody != null)
+                    {
+                        body = sanitizedBody;
+                    }
+                }
+
+                Dictionary<string, object> logContext = headers.Concat(body).ToDictionary(kv => kv.Key, kv => (object) kv.Value);
+                _logger.LogRequest(httpContext.Request, httpContext.Response, duration, logContext);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogCritical(exception, "Failed to track request");
+            }
+        }
+
+        /// <summary>
+        /// Extracts information from the given HTTP <paramref name="requestHeaders"/> to include in the request tracking context.
+        /// </summary>
+        /// <param name="requestHeaders">The headers of the current HTTP request.</param>
+        protected virtual IDictionary<string, StringValues> SanitizeRequestHeaders(IDictionary<string, StringValues> requestHeaders)
+        {
+            return requestHeaders.Where(header => _options.OmittedHeaderNames.Contains(header.Key) == false);
+        }
+
+        private static async Task<IDictionary<string, StringValues>> SanitizeRequestBodyAsync(Stream requestStream)
+        {
+            if (!requestStream.CanRead)
+            {
+                return new Dictionary<string, StringValues>();
+            }
+
+            if (requestStream.CanSeek)
+            {
+                if (requestStream.Position != 0)
+                {
+                    requestStream.Position = 0;
+                }
+            }
+            else
+            {
+                try
+                {
+                    if (requestStream.Position != 0)
+                    {
+                        return new Dictionary<string, StringValues>();
+                    }
+                }
+                catch
+                {
+                    return new Dictionary<string, StringValues>();
+                }
+            }
+
+            using (var reader = new StreamReader(requestStream))
+            {
+                string contents = await reader.ReadToEndAsync();
+                return new Dictionary<string, StringValues>
```

Comment:
This still feels a bit odd to me, think this one should just return the raw string and you can then move it to a dictionary if you want; but from a consumer perspective if you are overriding this you shouldn't have to care about this (that's our problem)

Reply:
Ah, OK, I thought that it was also a consumer plus to have the opportiunity to add additional entries, or even split your body into separate entries. But I'll update this then to just returning the raw body.